sql-to-dbsp-compiler/lib/sqllib/src/variant.rs

```diff
@@ -146,29 +147,237 @@ pub enum Variant {
+                }
+
+                Ok(Variant::Array(vec))
+            }
+
+            #[inline]
+            fn visit_map<V>(self, mut visitor: V) -> Result<Variant, V::Error>
+            where
+                V: MapAccess<'de>,
+            {
+                match visitor.next_key_seed(KeyClassifier)? {
+                    Some(KeyClass::Number) => {
+                        let number: NumberFromString = visitor.next_value()?;
+                        Ok(Variant::Decimal(number.value))
+                    }
+                    Some(KeyClass::Map(first_key)) => {
+                        let mut values = BTreeMap::new();
+
+                        values.insert(Variant::String(first_key), visitor.next_value()?);
+                        while let Some((key, value)) = visitor.next_entry::<String, Variant>()? {
+                            values.insert(Variant::String(key), value);
+                        }
+
+                        Ok(Variant::Map(values))
+                    }
+                    None => Ok(Variant::Map(BTreeMap::new())),
                 }
             }
         }
+
+        deserializer.deserialize_any(VariantVisitor)
+    }
+}
+
+struct KeyClassifier;
+
+enum KeyClass {
+    Map(String),
+    Number,
+}
+
+impl<'de> DeserializeSeed<'de> for KeyClassifier {
+    type Value = KeyClass;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_str(self)
+    }
+}
+
+// This is defined in serde_json, but not exported.
+const DECIMAL_KEY_TOKEN: &str = "$serde_json::private::Number";
+
+impl<'de> Visitor<'de> for KeyClassifier {
+    type Value = KeyClass;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a string key")
+    }
+
+    fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        match s {
+            DECIMAL_KEY_TOKEN => Ok(KeyClass::Number),
+            _ => Ok(KeyClass::Map(s.to_owned())),
+        }
+    }
+
+    fn visit_string<E>(self, s: String) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        match s.as_str() {
+            DECIMAL_KEY_TOKEN => Ok(KeyClass::Number),
+            _ => Ok(KeyClass::Map(s)),
+        }
+    }
+}
+
+pub struct NumberFromString {
+    pub value: Decimal,
+}
+
+impl<'de> de::Deserialize<'de> for NumberFromString {
+    fn deserialize<D>(deserializer: D) -> Result<NumberFromString, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        struct Visitor;
+
+        impl<'de> de::Visitor<'de> for Visitor {
+            type Value = NumberFromString;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("string containing a decimal number")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<NumberFromString, E>
```

Comment:
is this expected to consume the whole string?
because otherwise it's ambiguous

Reply:
yes, it should consume the entire string that represents the decimal number.